soc: nxp: imx95: Fix systick flooding by enabling cache earlier

PetervdPerk-NXP opened the pull request:
Systick is already enabled before late hook got called. In certain circumstances systick could be flooding the core and thus only handling systicks interrupts instead of executing code.

Enabling cache earlier in soc_init ensures that systick get handled quicker.

JiafeiPan reviewed (commented):
Whether could enable cache in soc_early_init_hook() ?  it happens more early than some device initialization (PRE_KERNEL_1 AND PRE_KERENL_2) ?

PetervdPerk-NXP commented:
> Whether could enable cache in soc_early_init_hook() ? it happens more early than some device initialization (PRE_KERNEL_1 AND PRE_KERENL_2) ?

Don't know the details but it happens after enabling spi devices. If an spi device needs to a lot of transactions then systick floods and system never boots on a xip flash system.

I just marched how its done on a imxrt target which also xip from flash

PetervdPerk-NXP committed:
soc: nxp: imx95: Fix systick flooding by enabling cache earlier

Systick is already enabled before late hook got called. In certain
circumstances systick could flooding the core and thus only handling
systicks. Enabling cache earlier using soc_early_init_hook ensures
that systick get handled quicker.

Signed-off-by: Peter van der Perk <peter.vanderperk@nxp.com>

PetervdPerk-NXP commented:
> > > Whether could enable cache in soc_early_init_hook() ? it happens more early than some device initialization (PRE_KERNEL_1 AND PRE_KERENL_2) ?
> > 
> > 
> > Don't know the details but it happens after enabling spi devices. If an spi device needs to a lot of transactions then systick floods and system never boots on a xip flash system.
> > I just matched how its done on a imxrt target which also xip from flash
> 
> If spi driver could cause the issue, so I am not sure whether other driver could also cause the same issue, so I think put it to be ealy init hook will be better, please have a try.

It's not a driver issue, it simply too much happening during init and thus when doing XIP from flash that seeking to new addresses is too slow without the instruction cache.

Using `soc_early_init_hook` solves the issue for me by enabling the instruction cache earlier thus reducing pressure on the XIP controller.
